simple-dns/src/dns/rdata/svcb.rs

```diff
@@ -0,0 +1,326 @@
+use std::{borrow::Cow, convert::TryInto};
+
+use crate::dns::{PacketPart, MAX_SVC_PARAM_VALUE_LENGTH};
+use crate::{CharacterString, Name};
+
+use super::RR;
+
+/// The SVCB DNS RR type is used to locate alternative endpoints for a service.
+/// [RFC 9460](https://datatracker.ietf.org/doc/html/rfc9460).
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub struct SVCB<'a> {
+    /// The priority of this record (relative to others, with lower values preferred).
+    ///
+    /// A value of 0 indicates AliasMode.
+    pub priority: u16,
+
+    /// The domain name of either the alias target (for AliasMode)
+    /// or the alternative endpoint (for ServiceMode).
+    pub target: Name<'a>,
+
+    /// A list of key=value pairs describing the alternative endpoint at `target`.
+    pub params: Vec<SvcParam<'a>>,
```

Comment:
If I understand correctly the [RFC](https://www.rfc-editor.org/rfc/rfc9460.html#section-2.2-5), this list needs to be ordered when building the wire format.
> SvcParamKeys SHALL appear in increasing numeric order.

To avoid generating invalid packets, I recommend a few changes around this:
- Implement `Ord, PartialOrd` for `SvcParam` based on the key
- Manually implement `Hash` to consider just the key
- Use a `BTreeSet` instead of a `Vec` 

Another option would be to iterate over the params collection before writing the wire format and check if the keys are in order.

Reply:
Implementing `Ord` (and thus `Eq`) based on key alone will make `SvcParam` with different values be treated as equal, that makes comparing two `SVCB` structures invalid :thinking: 

If we need to enforce the ordering `params` needs to become `BTreeMap<u16, Cow<'a, [u8]>>`. In order to keep the length checking of those SvcParamValue I think we need to also make `params` private, and move all the SvcParam-related methods into the `SVCB` structure itself.

Reply:
Clients MUST consider an RR malformed if:

    the end of the RDATA occurs within a SvcParam.
    SvcParamKeys are not in strictly increasing numeric order.
    the SvcParamValue for a SvcParamKey does not have the expected format.

    Note that the second condition implies that there are no duplicate SvcParamKeys.

There shouldn't be duplicated keys, according to the RFC.